client/src/main/java/org/apache/rocketmq/client/impl/ClientRemotingProcessor.java

```diff
@@ -213,4 +221,79 @@ private RemotingCommand consumeMessageDirectly(ChannelHandlerContext ctx,
+                String bornHost/*ip:port*/ = bornHostArr[bornHostArr.length - 1];
+                String[] host = bornHost.split(":");
+                if (host.length == 2)
+                    msg.setBornHost(new InetSocketAddress(host[0], Integer.parseInt(host[1])));
+            }
+
+            if (requestHeader.getStoreHost() != null) {
+                String[] storeHostArr = requestHeader.getStoreHost().split("/");
+                String storeHost = storeHostArr[storeHostArr.length - 1];
+                String[] host = storeHost.split(":");
+                if (host.length == 2)
+                    msg.setStoreHost(new InetSocketAddress(host[0], Integer.parseInt(host[1])));
+            }
+
+            byte[] body = request.getBody();
+            if ((requestHeader.getSysFlag() & MessageSysFlag.COMPRESSED_FLAG) == MessageSysFlag.COMPRESSED_FLAG) {
+                try {
+                    body = UtilAll.uncompress(body);
+                } catch (IOException e) {
+                    log.warn("err when uncompress constant", e);
+                }
+            }
+            msg.setBody(body);
+            msg.setFlag(requestHeader.getFlag());
+            MessageAccessor.setProperties(msg, MessageDecoder.string2messageProperties(requestHeader.getProperties()));
+            MessageAccessor.putProperty(msg, MessageConst.PROPERTY_REPLY_MESSAGE_ARRIVE_TIME, String.valueOf(receiveTime));
+            msg.setBornTimestamp(requestHeader.getBornTimestamp());
+            msg.setReconsumeTimes(requestHeader.getReconsumeTimes() == null ? 0 : requestHeader.getReconsumeTimes());
+            log.debug("receive reply message :{}", msg);
+
+            processReplyMessage(msg);
+
+            response.setCode(ResponseCode.SUCCESS);
+            response.setRemark(null);
+        } catch (Exception e) {
+            log.warn("unknown err when receiveReplyMsg", e);
+            response.setCode(ResponseCode.SYSTEM_ERROR);
+            response.setRemark("process reply message fail");
+        }
+        return response;
+    }
+
+    private void processReplyMessage(MessageExt replyMsg) {
+        final String correlationId = replyMsg.getUserProperty(MessageConst.PROPERTY_CORRELATION_ID);
+        final RequestResponseFuture requestResponseFuture = RequestFutureTable.getRequestFutureTable().get(correlationId);
+        if (requestResponseFuture != null) {
+            requestResponseFuture.putResponseMessage(replyMsg);
+
+            RequestFutureTable.getRequestFutureTable().remove(correlationId);
+
+            if (requestResponseFuture.getRequestCallback() != null) {
+                requestResponseFuture.getRequestCallback().onSuccess(replyMsg);
+            } else {
+                requestResponseFuture.putResponseMessage(replyMsg);
+            }
+        } else {
+            log.warn(String.format("receive reply message, but not matched any request, CorrelationId: %s", correlationId));
```

Comment:
It would be nice if can print the born host and msgId for the condition of receiving more than one replies because multi-consumer group subscribed to the same request topic.

Reply:
Is the msgId of request message or reply message?

Reply:
Eh, just a suggestion for tracing the reply whether in line with expectations, maybe the reply message be better.

Reply:
In this implement, ReplyMessageProcessor will push reply message to producer and then store the reply message, so that it has low latency without waiting store message. So the reply message pushed to producer has no msgId. Would it be better that store reply message first and then push reply message to producer so that reply message can get a msgId ?